src/models/transactions.ts

```diff
@@ -0,0 +1,100 @@
+import { connection, Schema, model, Model, Types } from "mongoose";
+import type { IAnyObject } from "../IAnyObject";
+import type { IJournal } from "./journals";
+
+export interface ITransaction {
+  _id: Types.ObjectId;
+  credit: number;
+  debit: number;
+  meta: IAnyObject;
+  datetime: Date;
+  account_path: string[];
+  accounts: string;
+  book: string;
+  memo: string;
+  _journal: Types.ObjectId | IJournal;
+  timestamp: Date;
+  voided: boolean;
+  void_reason?: string;
+  approved: boolean;
+  _original_journal?: Types.ObjectId | IJournal;
+}
+
+export const transactionSchema = new Schema<ITransaction>(
+  {
+    credit: Number,
+    debit: Number,
+    meta: Schema.Types.Mixed,
+    datetime: Date,
+    account_path: [String],
+    accounts: String,
+    book: String,
+    memo: String,
+    _journal: {
+      type: Schema.Types.ObjectId,
+      ref: "Medici_Journal",
+    },
+    timestamp: Date,
+    voided: {
+      type: Boolean,
+      default: false,
+    },
+    void_reason: String,
+    // The journal that this is voiding, if any
+    _original_journal: {
+      type: Schema.Types.ObjectId,
+      ref: "Medici_Journal",
+    },
+    approved: {
+      type: Boolean,
+      default: true,
+    },
+  },
+  { id: false, versionKey: false, timestamps: false }
+);
+
+export let transactionModel: Model<ITransaction>;
+
+let transactionSchemaKeys: Set<string> = new Set(
+  Object.keys(transactionSchema.paths)
+);
+
+export function isValidTransactionKey<T extends ITransaction = ITransaction>(
+  value: unknown
+): value is keyof T {
+  return typeof value === "string" && transactionSchemaKeys.has(value);
+}
+
+export function setTransactionSchema(schema: Schema, collection?: string) {
+  delete connection.models["Medici_Transaction"];
+
+  schema.index({ _journal: 1 });
+  schema.index({
+    accounts: 1,
+    book: 1,
+    approved: 1,
+    datetime: -1,
+    timestamp: -1,
+  });
+  schema.index({ "account_path.0": 1, book: 1, approved: 1 });
+  schema.index({
+    "account_path.0": 1,
```

Comment:
Usually the indizes should be sorted by cardinality of each item. See https://docs.mongodb.com/manual/core/index-compound/#prefixes

If we think, that Accounting are structured like 

"Expenses:Salary:Employer1"
"Expenses:Salary:Employer2"
"Expenses:Salary:Employer3"

Than the indizes should be in the reverse order, like `account_path.1` should be before `account_path.0` as the cardinality would be in the reverse order. But if you have Accounts structured like

"Employer1:Salary:Expenses"
"Employer2:Salary:Expenses"
"Employer3:Salary:Expenses"

than current indizes make sense.

But somehow I think it is actually the first case, as the README.md suggests that structure. Maybe we should determine what the most used account naming style is, and then recommend it as best practice. Maybe we should also add a syncIndices helper method as with setTransactionSchema you could add your own indizes and after you set the transactionSchema with setTransactionSchema you could use the syncIndizes-helper to enforce the indizes.

Maybe setTransactionSchema should have an Option parameter, were you can select if you want to add the default-indizes or not.

to Everyone:
Any thoughts?

Reply:
Added syncIndexes.

Reply:
Adding an index datetime and timestamp improves the performance of ledger noticable.

Reply:
Added documentation about cardinality.